packages/channel-provider/src/types.ts

```diff
@@ -120,11 +121,15 @@ export type MethodRequestType =
   | Call<'GetBudget', GetBudgetRequest>
   | Call<'CloseAndWithdraw', any>;
 
+const eventEmitter = new EventEmitter<NotificationType>();
+export type OnType = typeof eventEmitter.on;
+export type OffType = typeof eventEmitter.off;
+
 export interface ChannelProviderInterface {
   enable(url?: string): Promise<void>;
   send(request: MethodRequestType): Promise<MethodResponseType[MethodRequestType['method']]>;
-  on(event: string, callback: ListenerFn): void;
-  off(event: string, callback?: ListenerFn): void;
-  subscribe(subscriptionType: string, params?: any): Promise<string>;
-  unsubscribe(subscriptionId: string): Promise<boolean>;
+  on: OnType;
+  off: OffType;
+  // subscribe(subscriptionType: string, params?: any): Promise<string>;
+  // unsubscribe(subscriptionId: string): Promise<boolean>;
 }
```

Comment:
This captures the main idea: `callback` now has a type that depends on the type of `event`, meaning passing in the wrong type / mismatched of parameter to the callback will be a type error. E.g. `.on('BudgetUpdated', (channelResult: ChannelResult) => void)`.

Instead of explicitly constructing the conditional types, here we just let the EventEmitter class do that for us, by passing in `<NotificationType>` to its constructor and grabbing a type from the resulting instance.

`NotificationType` is a lookup from event name to params for that event.